hardware/mifare/mifare_classic.ksy

```diff
@@ -0,0 +1,243 @@
+meta:
+  id: mifare_classic
+  endian: le
+  file-extension: mfd
+  title: "Mifare Classic RFID tag dump"
+  license: BSD-2-Clause
+doc-ref: |
+  https://github.com/nfc-tools/libnfc
+  https://www.nxp.com/docs/en/data-sheet/MF1S70YYX_V1.pdf
+seq:
+  - id: sectors
+    type: sector(_index)
+    repeat: eos
+types:
+  key:
+    seq:
+      - id: key
+        size: 6
+  sector:
+    params:
+      - id: idx
+        type: u1
```

Comment:
The sole purpose of `idx` here is to make sure that before the body of very first sector, a `manufacturer` signature will appear. Why don't just move in outside of the sector, and get rid of this `idx` and whole params thing?

Reply:
Because it breaks incapsulation. We wanna access all the sectors data in a uniform way. I mean the 0th sector contains data too, and when we loop over sectors we wanna access the data in it too without handling this situation in our code. I expect the compiler be somewhen smart and featureful enough to optimize this stuff out.

Reply:
I don't think it has much to do with incapsulation. If, by definition, there could be only 1 `manufacturer` per `mifare_classic`, it makes total sense to have it as an attribute in `mifare_classic`.

Dealing with sectors and blocks might make some sense when one's writing a flasher app or something like that, but it does not matter if you're acting on a card dump. If they say that there are "sectors", but first "sector" is totally not like all others, it makes sense either to do a special type for that first "sector", or not define first "sector" as sector, but rather as a distinct list of attributes.

Reply:
>but first "sector" is totally not like all others

It is like all others, the only difference is manufacturer block (in fact it's just a block of data written on factory and locked, but there exist cards (so called "Mifare Zero") with that block unlocked).

Reply:
I won't push any strong opinions on this, but, generally, what you're dealing here is a card dump. There is no notion of "sectors" or "blocks" in a stream of bytes — you are free to read arbitrary locatons and define whatever data blocks you want. As far as I understand, this whole "sectors" and "blocks" thing is relevant if you are doing some card reader / writer software — and if you do, you generally don't care what's inside these sectors/blocks, you just use them as ready-made data blocks.

Reply:
>As far as I understand, this whole "sectors" and "blocks" thing is relevant if you are doing some card reader / writer software — and if you do, you generally don't care what's inside these sectors/blocks, you just use them as ready-made data blocks.

The logic of apps using nfc cards is closely tied to blocks, values and sectors since the card IC "enforce" (the crypto was broken, see mfcuk and mfoc) access conditions on block granularity. Modern apps just use cards as dumb storage using own digital signatures, but still write specific data on specific offsets.

>you are doing some card reader / writer software

All software working with the cards is sort of reader/writer software. Though most of them are not generic readers/writers but application-specific ones combining in a single app reading a card, decoding its data and showing it to a user. Card dumps are often incomplete because some keys may be unknown (and the tools mentioned are usually incompatible with the nfc modules used in android phones - the most widespread hardware for reading and writing cards).

Reply:
Ok, may be then we should start with some sort of application that does this "read → decode → show to user" routine and see how can we make KS useful for it? For example, if it only works on sector level, there is no point in having this `sectors` array at all.

Reply:
>there is no point in having this sectors array at all.

If I understand right, mfd format is just a sequence of raw sectors. The ones needeng to parse sectors or their parts separately can use the types defined here, as I have done in https://github.com/KOLANICH/MFCTool.py

Reply:
Tested on the mfd. Works.

Reply:
@GreyCat seems to be ready to be merged.